spacemouse-keys/hidInterface.cpp

```diff
@@ -4,25 +4,167 @@
+uint8_t countTransZeros = 0; // count how many times, the zero data has been sent
+uint8_t countRotZeros = 0;
+
+// Function to send translation and rotation data to the 3DConnexion software using the HID protocol
+// Three sets of data are sent: translation, rotation and key events
 // For each, a 16bit integer is split into two using bit shifting. The first is mangitude and the second is direction.
-void send_command(int16_t rx, int16_t ry, int16_t rz, int16_t x, int16_t y, int16_t z, uint8_t *keys, int debug)
+// With regard to the _hidReportDescriptor: the values should be within +/- 350.
+// returns true, if new data was sent
+// A HID report may be sent every HIDUPDATERATE_MS, which should be 8 ms
+bool send_command(int16_t rx, int16_t ry, int16_t rz, int16_t x, int16_t y, int16_t z, uint8_t *keys, int debug)
 {
-  uint8_t trans[6] = {(byte)(x & 0xFF), (byte)(x >> 8), (byte)(y & 0xFF), (byte)(y >> 8), (byte)(z & 0xFF), (byte)(z >> 8)};
-  HID().SendReport(1, trans, 6);
-  uint8_t rot[6] = {(byte)(rx & 0xFF), (byte)(rx >> 8), (byte)(ry & 0xFF), (byte)(ry >> 8), (byte)(rz & 0xFF), (byte)(rz >> 8)};
-  HID().SendReport(2, rot, 6);
+  static unsigned long now = 0; // time from millis()
+  now = millis();
+
+  static unsigned long lastHIDsentRep = now; // time from millis(), when the last HID report was sent
+
+  bool hasSentNewData = false; // this value will be returned
 
 #if (NUMKEYS > 0)
-  static uint8_t data[HIDMAXBUTTONS / 8];     // array to be sent over hid
+  static uint8_t keyData[HIDMAXBUTTONS / 8];     // key data to be sent via HID
+  static uint8_t prevKeyData[HIDMAXBUTTONS / 8]; // previous key data
+  prepareKeyBytes(keys, keyData, debug);         // sort the bytes from keys into the bits in keyData
+#endif
+
+  switch (nextState) // state machine
+  {
+  case ST_START:
+    // Evaluate everytime, without waiting for 8ms
+    if (countTransZeros < 3 || countRotZeros < 3 || (x != 0 || y != 0 || z != 0 || rx != 0 || ry != 0 || rz != 0))
+    {
+      // if one of the values is not zero,
+      // or not all zero data packages are sent (sent 3 of them)
+      // start sending data
+      nextState = ST_SENDTRANS;
+    }
+    else
+    {
+// if nothing is to be sent, check for keys. If no keys, don't change state
+#if (NUMKEYS > 0)
+      if (memcmp(keyData, prevKeyData, HIDMAXBUTTONS / 8) != 0)
+      // compare key data to previous key data
+      {
+        nextState = ST_SENDKEYS;
+      }
+#endif
+      if (nextState == ST_START && (now - lastHIDsentRep >= HIDUPDATERATE_MS))
+      {
+        // if we are not leaving the start state and
+        // we are waiting here for more than the update rate,
+        // keep the timestamp for the last sent package nearby
+        lastHIDsentRep = now - HIDUPDATERATE_MS;
+      }
+    }
+    break;
+  case ST_SENDTRANS:
+    // send translation data, if the 8 ms from the last hid report have past
+    if (now - lastHIDsentRep >= HIDUPDATERATE_MS)
```

Comment:
what do you think to extract this test `now - lastHIDsentRep >= HIDUPDATERATE_MS` in a method to give it a more meaningful name ?

Reply:
Agreed. IsNewHidReportDue(now) is part of my next commit.